src/context.rs

```diff
@@ -10,6 +10,211 @@ use crate::ffi::types::{c_uint, c_void, AlignedType};
+    /// Only a bit or two per signing operation is needed; if you have any entropy at all,
+    /// you should provide it, even if you can't provide 32 random bytes.
+    pub fn with_raw_global_context<T, F: FnOnce(NonNull<ffi::Context>) -> T>(
+        f: F,
+        rerandomize_seed: Option<&[u8; 32]>,
+    ) -> T {
+        assert!(
+            unsafe {
+                ffi::secp256k1_context_preallocated_size(AllPreallocated::FLAGS)
+                    <= core::mem::size_of::<[AlignedType; MAX_PREALLOC_SIZE]>()
+            },
+            "prealloc size exceeds our guessed compile-time upper bound"
+        );
+
+        // Our function may be expensive, so before calling it, we copy the global
+        // context into this local buffer on the stack. Then we can release it,
+        // allowing other callers to use it simultaneously.
+        let mut store = [AlignedType::ZERO; MAX_PREALLOC_SIZE];
+        let buf = NonNull::new(store.as_mut_ptr() as *mut c_void).unwrap();
+
+        let ctx = match SECP256K1.try_lock() {
+            None => unsafe {
+                // If we can't get the lock, just do everything on the stack.
+                ffi::secp256k1_context_preallocated_create(buf, AllPreallocated::FLAGS)
+            },
+            Some(ref mut guard) => unsafe {
+                // If we *can* get the lock, use it and update it.
+                let (ref mut store, ref mut ctx) = **guard;
+                let global_ctx = ctx.get_or_insert_with(|| {
+                    let buf = NonNull::new(store.as_mut_ptr() as *mut c_void).unwrap();
+                    ffi::secp256k1_context_preallocated_create(buf, AllPreallocated::FLAGS)
+                });
+                ffi::secp256k1_context_preallocated_clone(global_ctx.as_ptr(), buf)
+            },
+        };
+        // The lock is now dropped. Call the function.
+        let ret = f(ctx);
+        // ...then rerandomize the local copy, and try to replace the global one
+        // with this. There are three cases for how this can work:
+        //
+        //     1. In the happy path, we succeeded in getting the lock above, have
+        //        a copy of the global context, are rerandomizing and storing it.
+        //        Great.
+        //     2. Same as above, except that another thread is doing the same thing
+        //        in parallel. Now we both have copies that we're rerandomizing, and
+        //        both will try to store it. One of us will clobber the other, wasting
+        //        work but otherwise not causing any problems.
+        //     3. If we -failed- to get the lock above, we are rerandomizing a fresh
+        //        copy of the context object. This may "undo" previous rerandomization.
+        //        In theory if an attacker is able to reliably and repeatedly trigger
+        //        this situation, they will have defeated the rerandomization. Since
+        //        this is a defense-in-depth measure, we will accept this.
```

Comment:
Shouldn't we just remember that we have a fresh copy and rerandomization is pointless?

Reply:
In the current code, even if we have a fresh copy, we might replace the global one with the fresh copy in the end, so we should rerandomize it.

Maybe we shouldn't do this? I could go either way.

Reply:
Ok, switched to this logic (and updated the comment accordingly).